signac/core/collection_api.py

```diff
@@ -0,0 +1,446 @@
+                    "keys may not contain dots ('.'): {}".format(key))
+            else:
+                return key
+        elif isinstance(key, cls.VALID_KEY_TYPES):
+            return cls._validate_key(str(key))
+        else:
+            from ..errors import KeyTypeError
+            raise KeyTypeError(
+                "keys must be str, int, bool or None, not {}".format(type(key).__name__))
+
+    def __delitem__(self, item):
+        self.load()
+        del self._data[item]
+        self.sync()
+
+    def __setitem__(self, key, value):
+        self.load()
+        with self._suspend_sync():
+            self._data[self._validate_key(key)] = self.from_base(value, parent=self)
+        self.sync()
+
+    def __getitem__(self, key):
+        self.load()
+        return self._data[key]
+
+    def __iter__(self):
+        self.load()
+        return iter(self._data)
+
+    def __call__(self):
+        self.load()
+        return self.to_base()
+
+    def __eq__(self, other):
+        if isinstance(other, type(self)):
+            return self() == other()
+        else:
+            return self() == other
+
+    def __len__(self):
+        self.load()
+        return len(self._data)
+
+    def keys(self):
+        self.load()
+        return self._data.keys()
+
+    def values(self):
+        self.load()
+        return self.to_base().values()
+
+    def items(self):
+        self.load()
+        return self.to_base().items()
+
+    def get(self, key, default=None):
+        self.load()
+        return self._data.get(key, default)
+
+    def pop(self, key, default=None):
+        self.load()
+        ret = self._data.pop(key, default)
+        self.sync()
+        return ret
+
+    def popitem(self, key, default=None):
+        self.load()
+        ret = self._data.pop(key, default)
+        self.sync()
+        return ret
+
+    def clear(self):
+        self._data = {}
+        self.sync()
+
+    def update(self, mapping):
+        self.load()
+        with self._suspend_sync():
+            for key, value in mapping.items():
+                self[key] = self.from_base(value, self)
+        self.sync()
+
+    def setdefault(self, key, default=None):
+        self.load()
+        with self._suspend_sync():
+            ret = self._data.setdefault(key, self.from_base(default, self))
+        self.sync()
+        return ret
+
+
+class SyncedDict(_SyncedDict):
+
+    _PROTECTED_KEYS = ('_data', '_suspend_sync_', '_load', '_sync', '_parent')
+
+    def __getattr__(self, name):
+        try:
+            return super(SyncedDict, self).__getattribute__(name)
+        except AttributeError:
+            if name.startswith('__'):
+                raise
+            try:
+                return self.__getitem__(name)
+            except KeyError as e:
+                raise AttributeError(e)
+
+    def __setattr__(self, key, value):
+        try:
+            super(SyncedDict, self).__getattribute__('_data')
```

Comment:
When would this fail? Seems like `self._data` always exists after construction?

Reply:
It fails while creating an instance of a child of `SyncedDict`(before the data is assigned). I don't need to add the data members of the child in `_Protected_key` as long as they are assigned before the call of `super().__init__()`.